### PR TITLE
[Non-urgent] AG-2017: normalize null values util function

### PR DESCRIPTION
**Pull request by jaclynbeck-sage**
This PR addresses some of [AG-2017](https://sagebionetworks.jira.com/browse/AG-2017) by creating a util function that normalizes null values in a data frame.

### Problem

When pandas reads a CSV, all missing values are `NaN`. This results in several issues:
- The transforms have been selectively using `.fillna()` on some fields but not others, or using `.fillna("")` on a whole data frame, which affects numeric columns and boolean columns and requires casting those back to their original data type
- Some transforms do these `.fillna()` operations inconsistently on the same data
- `NaN` and `None` are treated differently. For example, `bool(NaN)` returns `True` while `bool(None)` returns `False`
- `NaN` values need to be converted to `None` when writing out to json anyway
- On data frame merge, missing values created by non-matching data are filled with `NaN`, regardless of column type. This can result in columns with both `NaN` and `None` values.

### Solution

Created a utility function, `normalize_null_values`, that does the following on a given data frame:
1. By default, all `np.NaN` values are set to `None`
2. If the (optional) `boolean_columns` argument contains a list of column names, those columns will have all `NaN` and `None` values replaced with `False`, and the column type is set to `boolean`. 
3. If the (optional) `empty_string_columns` argument contains a list of column names, those columns will have all `NaN` and `None` values replaced with empty strings (""). 

This function provides a more consistent (and more compact) way to fill NA values in a data frame.

### Changes in this PR

- Added the `normalize_null_values` function and a test suite to `utils.py`
- Updated **utils/nest_fields** function:
    - using `normalize_null_values` does vectorized replacement of np.NaN with None outside the lambda instead of inside it. Unsure if this makes a difference but it seems cleaner.
- Updated **model_ad_transform_utils**:
    - `process_genetic_info`: normalized the merged allele_info and human_gene_symbol_map data frame
    - `build_transcriptomic_url`: code adjusted to assume that any input data frames have been normalized prior to calling the function (boolean NA data filled with False, other column NA data replaced with None). 
        - All transforms that call this function have been updated to normalize their null values first.
    - Updated `build_transcriptomic_url` tests to reflect empty value change from "" to None
- Updated **model_details** transform:
    - merged model_info_df is normalized (boolean columns filled with False, some string columns filled with "", all others filled with None)
    - removed unnecessary `.fillna("")` from allele_info and human_transgene_allele_map
    - removed test for missing data in allele_info for this transform: none of the fields tested should be empty. Instead, this is a candidate for checking non-empty with column rules. 
- Updated **model_overview** transform:
    - merged model_info_df is normalized (boolean columns filled with False, all others filled with None)
    - code cleaned up now that we no longer need to check for "" and NaN or cast values to booleans
    - updated tests to reflect empty URL value change from "" to None

#### Tests for normalize_null_values

- correctly normalizes with no `boolean_columns` or `empty_string_column` arguments specified
- correctly normalizes with only `boolean_columns` specified
- correctly normalizes with only `empty_string_column`  specified
- correctly normalizes with both `boolean_columns` and `empty_string_column` arguments specified
- empty data frames are unaltered
- fails if a column listed in `boolean_columns` or `empty_string_column` doesn't exist in the data frame
- fails if a column is listed in both `boolean_columns` and `empty_string_column`
- fails if the `df` argument is not a pd.DataFrame
- fails if the `boolean_columns` or `empty_string_column` argument is not a list

### Not addressed in this PR

No other transforms aside from `model_details` and `model_overview` have been changed to use `normalize_null_values` (yet). Some of the other transforms need more cleanup and it made this PR too large. They will be addressed in a separate PR(s).

[AG-2017]: https://sagebionetworks.jira.com/browse/AG-2017?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/agoradatatools/etl/transform/model_details.py [hunk 1]
```diff
@@ -114,8 +117,11 @@ def transform_model_details(
     check_required_datasets_and_columns(datasets, required_input)
 
     # Load and prepare datasets
-    allele_info_df = datasets["allele_info"].fillna("")
-    human_transgene_allele_map_df = datasets["human_transgene_allele_map"].fillna("")
+    allele_info_df = normalize_null_values(
+        datasets["allele_info"],
+        empty_string_columns=["gene_ensembl_id", "modified_gene", "allele"],
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py]: This is here so tests pass. In the real data, none of these fields are empty, and shouldn't be. This is a column rule that can be added when Beatriz adds column checking to this transform.

File: tests/transform/test_model_details.py [hunk 2]
```diff
@@ -99,18 +99,6 @@ class TestTransformModelDetails:
             },
             "model_details_transform_no_human_match_output.json",
         ),
-        (
```
Comment (jaclynbeck-sage) [inline tests/transform/test_model_details.py:102]: This test case was removed because it tested for missing data in fields that should never be empty. I noticed this when normalizing null values in these fields broke the tests but not the actual data processing pipeline. 

Missing values in these fields should get added as a check in Beatriz's new column rules code instead.

File: src/agoradatatools/etl/transform/model_details.py [hunk 3]
```diff
@@ -127,15 +128,12 @@ def transform_model_details(
...
         on="name",
         validate="one_to_one",
-    ).fillna(
-        {
-            "transcriptomics": False,
-            "disease_correlation": False,
-            "rrid": "",
-            "alzforum_id": "",
-            "matched_controls": "",
-            "aliases": "",
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/model_details.py:137]: Aliases omitted from normalization call below, it is fine as None instead of ""

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py [hunk 4]
```diff
@@ -107,28 +111,28 @@ def build_transcriptomics_url(model_row: pd.Series) -> Union[str, None]:
 
     Args:
         model_row (pd.Series): A single row from the model_info data frame, which must contain columns "name",
-            "gene_expression", "url_categories_value", and "url_models_value". The latter two columns may be blank or
-            contain strings. "gene_expression" must be True, False, or None (which is interpreted as False).
+            "gene_expression", "url_categories_value", and "url_models_value". The latter two columns may be None or
+            contain strings. "gene_expression" must be True or False. It is assumed that normalize_null_values has
+            already been called on this data so that all missing values used in this function are None, not NA or empty
+            strings.
```
[the remark this answers is not in the document]
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:125]: Fixed in next commit

File: src/agoradatatools/etl/utils.py [hunk 5]
```diff
@@ -435,6 +432,84 @@ def extract_age_numeric(age: str) -> Union[int, None]:
     return int(match.group(1)) if match else None
 
 
+def normalize_null_values(
+    df: pd.DataFrame,
+    boolean_columns: list[str] = None,
+    empty_string_columns: list[str] = None,
```
[the remark this answers is not in the document]
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py]: Fixed in next commit

File: tests/test_utils.py [hunk 6]
```diff
@@ -923,6 +923,222 @@ def test_extract_age_numeric(self, input_age, expected):
...
+                "string1": ["abc", None, "", np.nan],
+                "string2": [np.nan] * 4,
+                "numeric1": [123, np.nan, np.nan, np.nan],
+                "numeric2": [np.nan] * 4,
+                "extra1": ["abc", "def", None, np.nan],
+                "extra2": [np.nan] * 4,
+            }
+        )
+
+    def _build_expected_output(
```
Comment (linglp) [inline tests/test_utils.py]: I think it might be better to just build all the expected dataframe in each individual test rather than use this helper function `_build_expected_output`. There's no test to test this helper function, and if there's a bug in _build_expected_output, the tests could pass while asserting the wrong thing.

File: src/agoradatatools/etl/utils.py [hunk 7]
```diff
@@ -202,12 +202,9 @@ def nest_fields(
         pd.DataFrame: New DataFrame with grouping column(s) and a column containing nested dictionaries
     """
     nested = (
-        df.groupby(grouping)
-        .apply(
-            lambda row: row.replace({np.nan: None})
-            .drop(columns=drop_columns)
-            .to_dict("records")
-        )
+        normalize_null_values(df)
```
Comment (linglp) [inline src/agoradatatools/etl/utils.py:359]: I checked out the test for `nest_fields`, but I am not seeing any null values in the test dataframe. Should we add something like: a DataFrame with NaN values in some columns, and the expected nested dicts should have None (not NaN) for those fields? That test would specifically cover the behavior that `normalize_null_values` is there to provide.

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py [hunk 8]
```diff
@@ -48,6 +50,8 @@ def process_genetic_info(
         how="left",
     )
 
+    merged_df = normalize_null_values(merged_df)
```
Comment (linglp) [inline src/agoradatatools/etl/transform/model_ad_transform_utils.py:61]: Based on the logic of the current code, if human_ensembl_id is "", then `pd.notna("")` is True, so the current code would set ensembl_gene_id to "". Is this an expected behavior?

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py [hunk 9]
```diff
@@ -10,6 +10,8 @@
 from typing import Any, Dict, List, Union
 import pandas as pd
 
+from agoradatatools.etl.utils import normalize_null_values
+
 
 def process_genetic_info(
```
Comment (linglp) [inline src/agoradatatools/etl/transform/model_ad_transform_utils.py:16]: I read the test for this funciton too, and it seems like there's no test to ensure that `NaN` values get converted to None. Can we add that?

File: src/agoradatatools/etl/transform/model_details.py [hunk 10]
```diff
@@ -117,8 +118,8 @@ def transform_model_details(
     check_required_datasets_and_columns(datasets, required_input)
 
     # Load and prepare datasets
-    allele_info_df = datasets["allele_info"].fillna("")
```
Comment (linglp) [inline src/agoradatatools/etl/transform/model_details.py:120]: I am not sure if this behavior change is intentional, so I just want to highlight this one here: previously, `.fillna("")` converts all the NaN values to "" in both human_transgene_allele_map_df and allele_info_df. Later, both data frames got passed into `process_genetic_info`. Even though there's `normalize_null_values` in `process_genetic_info`, the behavior is not exactly the same. `normalize_null_values` would convert the NaN to None while previously `NaN` gets converted to empty string.

Comment (linglp) [reply src/agoradatatools/etl/transform/model_details.py:120]: I also can't find tests for `allele_info` or `human_transgene_allele_map` that contains any blank/missing values. Is this tested anywhere?

**Review by linglp — COMMENTED** [review]
Hi @jaclynbeck-sage ! I think overall it looks good, and I just have some comments about the test. I noticed that there are some behavior changes without test coverage. Could you take a look? Thank you!

File: src/agoradatatools/etl/utils.py (hunk 7 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py:359]: Good catch!

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py (hunk 9 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:16]: Will do!

File: src/agoradatatools/etl/transform/model_details.py (hunk 10 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_details.py:120]: I'll double-check on this. I know when I ran the pipeline in full, removing the .fillna("") had no effect on the output of this transform. However that's all good data, I don't think I checked or tested for bad/non-ideal input.

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py (hunk 8 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:61]: I don't think I thought of this. When read from a file, missing Ensembl IDs are NaN and would never be an empty string unless we deliberately .fillna(""). I'll sift though the code and tests and see if it's necessary to add something for "" IDs.

File: tests/test_utils.py (hunk 6 above)
Comment (jaclynbeck-sage) [reply tests/test_utils.py]: Good point!

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py (hunk 9 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:16]: You've uncovered a bug! When one of the merge column values is missing, the row gets silently dropped. Let's assume that the data is pre-validated to have non-missing values in the relevant columns by using Beatriz's new column rule code. I'm keeping track of where we need to add this in a running doc. 

I've added a passing test for normalizing missing values in non-essential columns. I've also added a comment to the function description specifying that this function assumes no missing values in specific columns.

File: src/agoradatatools/etl/transform/model_details.py (hunk 10 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_details.py:120]: I would say, having NAs filled with "" is actually undesired behavior and can result in rows post-merge with empty gene names and/or mgi_allele_ids, which we don't want. So I think letting them be None instead is better. However see my note about us pre-validating that the data has no missing values.

There is one file-based test for a different transform (model_overview) that has blank items in allele_info (but not human_transgene_allele_map) that don't mess up the merge. The test is ID'ed as "Pass with missing data in some fields". I don't see it being tested in this transform or the other transform that uses these data sets (disease_correlation). I've made a note to add tests for this when we add column validation.

File: src/agoradatatools/etl/utils.py (hunk 7 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py:359]: Test added.

File: tests/test_utils.py (hunk 6 above)
Comment (jaclynbeck-sage) [reply tests/test_utils.py]: Fixed in next commit.

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py (hunk 8 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:61]: I think it's not necessary to check for this in this function. We can use the not_empty column rule to validate that there aren't any "" ensembl IDs.

File: src/agoradatatools/etl/utils.py [hunk 11]
```diff
@@ -435,6 +432,84 @@ def extract_age_numeric(age: str) -> Union[int, None]:
...
+    if overlaps:
+        raise ValueError(
+            f"Columns {sorted(overlaps)} appear in both the boolean_columns and empty_string_columns lists."
+        )
+
+    df = df.copy()
+
+    for col in boolean_columns:
+        df[col] = df[col].fillna(False).astype(bool)
+
```
[the remark this answers is not in the document]
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py:749]: Not going to worry about this. We are going to assume that boolean columns are read correctly by pandas.

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py [hunk 12]
```diff
@@ -20,6 +22,13 @@ def process_genetic_info(
     replace the ensembl_id with the human one. Each model's alleles are processed independently.
     Multiple entries are preserved for different alleles of the same gene.
 
+    This function assumes the input data has been pre-validated so that:
+        1. human_transgene_allele_map_df does not have any missing/NaN/None values in any column
+        2. model_alleles does not have any missing/NaN/None values in the "modified_gene" or "mgi_allele_id" columns,
+            which are needed for the merge. Missing values in other columns of model_alleles are allowed, and will be
+            normalized to None.
+    This is to ensure that rows are not silently dropped during the merge due to missing values in the merge columns.
```
[the remark this answers is not in the document]
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py]: Fixed in next commit.

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py [hunk 13]
```diff
@@ -92,7 +103,7 @@ def build_transcriptomics_url(model_row: pd.Series) -> Union[str, None]:
     two (or more) model names to the "models=..." part of the string.
 
     The exact values that should go in "categories=..." and "models=..." are pulled from columns in model_row. If the
-    url_categories_value is "", the "categories=..." string is not added. If the url_models_value is "", the
+    url_categories_value is None, the "categories=..." string is not added. If the url_models_value is None, the
     "models=..." string defaults to "models=<model_name>".
```
[the remark this answers is not in the document]
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:109]: Fixed in next commit.

File: tests/transform/test_model_ad_transform_utils.py [hunk 14]
```diff
@@ -306,7 +363,7 @@ def test_build_transcriptomics_url_missing_field(
 
         # Special case: model["name"] never gets used unless we set the url_models_value to empty
```
Comment (linglp) [inline tests/transform/test_model_ad_transform_utils.py]: Nit: test_build_transcriptomics_url_missing_field can be removed. The required columns are already validated upstream by check_required_datasets_and_columns, so a missing key can't reach this function in practice. The test only verifies that Python raises KeyError on a missing dict lookup, which is a built-in behavior that doesn't need testing.

File: tests/transform/test_model_ad_transform_utils.py [hunk 15]
```diff
@@ -271,13 +328,13 @@ def test_build_transcriptomics_url_default_category(
...
+        ids=["Pass with empty string value", "Pass with None value"],
     )
     def test_build_transcriptomics_url_default_models(
         self, empty_val: str, url_test_model: pd.Series
     ) -> None:
         """
-        The function should treat both "" and None/NA as empty values and have just the model name in the URL
+        The function should treat both "" and None as empty values and have just the model name in the URL
         """
         url_test_model["url_models_value"] = empty_val
```
Comment (linglp) [inline tests/transform/test_model_ad_transform_utils.py]: Every test in this class mutates the fixture directly. It works because pytest re-creates the fixture for each test, but this is in general not a good pracitce. I think a cleaner approach is just to create the input inline: 

```
@pytest.mark.parametrize(
    "empty_val",
    ["", None],
    ids=["Pass with empty string value", "Pass with None value"],
)
def test_build_transcriptomics_url_default_category(self, empty_val: str) -> None:
    """
    The function should treat both "" and None as empty values and not have a "categories=..." in the url
    """
    model = pd.Series(
        {
            "name": "Model",
            "url_categories_value": empty_val,
            "url_models_value": "model1,model2",
            "transcriptomics": True,
        }
    )
    url = build_transcriptomics_url(model)
    assert url == "comparison/expression?models=model1,model2"

```

File: src/agoradatatools/etl/transform/model_ad_transform_utils.py (hunk 9 above)
Comment (linglp) [reply src/agoradatatools/etl/transform/model_ad_transform_utils.py:16]: Okay that makes sense. Thanks for looking into this, @jaclynbeck-sage

File: src/agoradatatools/etl/transform/model_details.py (hunk 10 above)
Comment (linglp) [reply src/agoradatatools/etl/transform/model_details.py:120]: Great. Thanks for the follow up! Your note about pre-validating the data makes sense.

File: tests/transform/test_model_ad_transform_utils.py [hunk 16]
```diff
@@ -208,6 +208,63 @@ def test_process_genetic_info_case_insensitive_mapping(self) -> None:
...
+                "human_ensembl_id": ["ENSG00000123456", "ENSG00000987654"],
+            }
+        )
+
+        # The third gene does not exist in the human transgene mapping, so its Ensembl ID will not get overwritten in
+        # the output and the missing value will show up.
+        model_alleles = pd.DataFrame(
+            {
+                "modified_gene": ["App", "Mapt", "Psen1"],
+                "gene_ensembl_id": ["ENSMUSG00000011111", "ENSMUSG00000022222", np.NaN],
```
Comment (linglp) [inline tests/transform/test_model_ad_transform_utils.py]: See numpy release note here: https://numpy.org/devdocs/release/2.0.0-notes.html
I think we should avoid using `np.NaN` and switch to use `np.nan` in the code base since np.NaN will be deprecated in 2.0.0 of numpy

File: tests/test_utils.py [hunk 17]
```diff
@@ -539,6 +539,39 @@ def test_nest_fields_singlerow_nested_list_false(self):
         )
         assert list(nested_df["e"]) == expected_column_e
 
+    def test_nest_fields_normalizes_null_values(self) -> None:
+        df_with_nulls = pd.DataFrame(
+            {
+                "a": ["group_1", "group_1", "group_2"],
+                "b": ["1", np.NaN, "1"],
```
Comment (linglp) [inline tests/test_utils.py]: same here. Replace `np.NaN` with `np.nan`

File: src/agoradatatools/etl/utils.py [hunk 18]
```diff
@@ -435,6 +432,84 @@ def extract_age_numeric(age: str) -> Union[int, None]:
...
+    for col in empty_string_columns:
+        df[col] = df[col].fillna("")
+
+    # Replace any remaining NaN values with None
+    return df.replace(np.nan, None)
```
Comment (linglp) [inline src/agoradatatools/etl/utils.py]: Hmm just in case, also handle pd.NA: 
```
return df.replace({np.nan: None, pd.NA: None})
```

**Review by linglp — APPROVED** [review]
Hi @jaclynbeck-sage ! Thanks for your hard work! One main thing to note is that we should avoid mutating the fixture directly in every test. Mutations like this make the tests pretty fragile because other tests could easily break. Otherwise, it looks good. I could approve the PR now!

File: src/agoradatatools/etl/utils.py (hunk 18 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py]: Good catch, fixed in next commit.

File: tests/transform/test_model_ad_transform_utils.py (hunk 16 above)
Comment (jaclynbeck-sage) [reply tests/transform/test_model_ad_transform_utils.py]: Fixed in next commit.

File: src/agoradatatools/etl/utils.py (hunk 18 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py]: I also added some pd.NA values to the test input data frame to make sure this gets tested.

File: tests/transform/test_model_ad_transform_utils.py (hunk 15 above)
Comment (jaclynbeck-sage) [reply tests/transform/test_model_ad_transform_utils.py]: Fixed in next commit.

File: tests/transform/test_model_ad_transform_utils.py (hunk 14 above)
Comment (jaclynbeck-sage) [reply tests/transform/test_model_ad_transform_utils.py]: Fixed in next commit.

File: tests/transform/test_model_ad_transform_utils.py [hunk 19]
```diff
@@ -217,101 +274,84 @@ class TestBuildTranscriptomicsUrl:
...
+                "url_models_value": empty_val,
+                "transcriptomics": True,
+            }
+        )
 
-        url = build_transcriptomics_url(url_test_model)
+        url = build_transcriptomics_url(model)
         assert url == "comparison/expression?categories=category_string&models=Model"
 
-    @pytest.mark.parametrize(
```
Comment (jaclynbeck-sage) [inline tests/transform/test_model_ad_transform_utils.py:288]: Test unnecessary, removed per Lingling's comment.

**Comment by jaclynbeck-sage** [comment]
I think I've addressed all of @linglp 's comments, but leaving open awhile longer so @beatrizsaldana has a chance to review.

File: src/agoradatatools/etl/utils.py [hunk 20]
```diff
@@ -435,6 +432,84 @@ def extract_age_numeric(age: str) -> Union[int, None]:
...
+) -> pd.DataFrame:
+    """
+    Normalize null values in a DataFrame by replacing NaN or None values with False, empty strings, or None, depending
+    on the column specification. This function is necessary because pandas fills empty values with NaN when it reads
+    from a csv file, and having NaN in boolean and string columns instead of None can produce unexpected behavior.
+    There are also several transforms that benefit from using empty strings instead of None, so this function provides
+    that option too.
+
+    Boolean columns will have NaN/None values replaced with False, and string columns will have NaN/None values replaced
+    with "". After that, all columns left over will have their NaN values replaced with None.
```
Comment (beatrizsaldana) [inline src/agoradatatools/etl/utils.py]: NIT: I am not sure if I am reading this wrong, but the docstring makes me think that all string columns become "", but only empty_string_columns do; everything else becomes None via the final replace. Would you be able to clarify the docstring a bit to prevent this kind of minor confusion?

**Review by beatrizsaldana — APPROVED** [review]
Looks great! Left a minor nit, not a blocker so I'm approving. Thank you!!

File: src/agoradatatools/etl/utils.py (hunk 20 above)
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/utils.py]: Fixed, thanks for catching this!!
